### ISSUE TITLE
AutoPurge Not Working

### ISSUE THREAD

**Issue by lon-yang** [31z4/zookeeper-docker#30]
I'm using the image `zookeeper:3.4.11` to run container. And I have configured the `autoPurge` config. But it's not working, the `snapshot` and `log` file not deleteing. How can i fix this?

```conf
autopurge.purgeInterval=24
autopurge.snapRetainCount=10
```

**Comment by 31z4** [comment]
Hello @lon-yang Are you sure that the config is being correcly read by Zookeeper? Also, how do you pass this configuration to a container?

**Comment by lon-yang** [comment]
I'm using the volume binding. And the other option of this config is correct.

**Comment by 31z4** [comment]
@lon-yang could you please provide more details on how did you test it? Are you sure it's working as you expect without Docker?

**Comment by lon-yang** [comment]
I have another cluster with the same autopurge config without docker and it's working. In the dataLog's dir, the snap files count less than 30. But in the docker, the file count is more than 500 with one week later.


Best regards
--------------------------
Lon Yang
Email: dove2580@live.com<mailto:dove2580@live.com>

Elisey Zanko <notifications@github.com> 于 2017年12月9日 下午5:34写道：


@lon-yang<https://github.com/lon-yang> could you please provide more details on how did you test it? Are you sure it's working as you expect without Docker?

―

**Comment by 31z4** [comment]
@lon-yang please run `echo conf | nc localhost $ZOO_PORT` within the running container to make sure that the purge related configuration variables are in place and correct.

**Comment by lon-yang** [comment]
`echo conf` doesn't have the `autopurge` config, even the normal cluster with autopurge working

**Comment by rdouyere** [comment]
We have the same behaviour with Auto Purge not working on dockerized 3.4.11. Logs at startup seems to reflect our config:
```
2018-01-15 14:41:45,389 [myid:2] - INFO  [main:DatadirCleanupManager@78] - autopurge.snapRetainCount set to 3
2018-01-15 14:41:45,389 [myid:2] - INFO  [main:DatadirCleanupManager@79] - autopurge.purgeInterval set to 1
```

Attempt to use `zkCleanup.sh` fails as well. We did manage to perform a purge only by doing from inside the container:
```
java -cp zookeeper-3.4.11.jar:lib/slf4j-log4j1.2.16.jar:lib/slf4j-api-1.6.1.jar:lib:conf org.apache.zookeeper.server.PurgeTxnLog /data /datalog -n 4
```

**Comment by 31z4** [comment]
@rdouyere did you get any error when runnng `zkCleanup.sh`? Could you please paste it here if so?

**Comment by rdouyere** [comment]
If we do `zkCleanup.sh -n 3` from inside the container it returns instantly with no message.

**Comment by 31z4** [comment]
That's odd. My suggestion is to create an issue in the [upstream](https://issues.apache.org/jira/projects/ZOOKEEPER/issues). I have no idead how to troubleshoot this 😞

**Comment by rdouyere** [comment]
Ok thanks.

**Comment by marcosmoreno** [comment]
Did someone created the ticket in the upstream? I would do if not.

We are hitting the same problem, and forced to use a cron job that deletes the logs as follows:

    docker exec -it projectname_zookeeper_1 find /datalog/version-2 -type f -mtime +5 -name 'log.*' -exec rm -- '{}' \;

This is ugly and can even have unpredictable results, as it is not recommended to delete the files like that, but the zkCleanup.sh script does not work for us either, and data accumulates in Gb. Does anyone knows if the autopurge uses the zkCleanup script? Any workaround so far?

**Comment by 31z4** [comment]
@marcosmoreno please see [above](https://github.com/apache/zookeeper/pull/475).

**Comment by 31z4** [comment]
Fixed in `3.4.12`.

**Comment by ShubhS9** [comment]
Was only the zkCleanup fixed in 3.4.12 or the original issue where autopurging wasn't working was also fixed because I see the issue still happening in docker for the official docker image 3.4.14.

**Comment by kaustubhchoudhury-git** [comment]
The create_config() function of start-zookeeper script(for kubernetes) has a typo as follows,
`echo "autopurge.purgeInteval=$PURGE_INTERVAL" >> $CONFIG_FILE`
Change it to `autopurge.purgeInterval`
Because of the typo the actual `autopurge.purgeInterval` was never set and defaulted to 0. Resulted in `Purge task is not scheduled` in the zookeeper logs
